phoenix-core/src/main/java/org/apache/phoenix/index/GlobalIndexChecker.java

```diff
@@ -0,0 +1,366 @@
+                deleteRowIfAgedEnough(indexRowKey, ts);
+                row.clear();
+                return row;
+            }
+            // Scan the newly build row
+            indexScan.withStartRow(indexRowKey, true);
+            indexScan.withStopRow(indexRowKey, true);
+            RegionScanner regionScanner = region.getScanner(indexScan);
+            regionScanner.next(row);
+            regionScanner.close();
+            if (!row.isEmpty() && !verifyRowAndRemoveVerifiedCell(row)) {
+                deleteRowIfAgedEnough(indexRowKey, ts);
+                row.clear();
+            }
+            return row;
+        }
+
+        private boolean isVerifiedCell(Cell cell) {
+            return Bytes.compareTo(cell.getFamilyArray(), cell.getFamilyOffset(), cell.getFamilyLength(),
+                    emptyCF, 0, emptyCF.length) == 0 &&
+                    Bytes.compareTo(cell.getQualifierArray(), cell.getQualifierOffset(), cell.getQualifierLength(),
+                            emptyCQ, 0, emptyCQ.length) == 0;
+        }
+
+        private boolean verifyRow(byte[] rowKey, long ts) throws IOException {
+            Get get = new Get(rowKey);
+            get.setTimeRange(minTimestamp, maxTimestamp);
+            get.addColumn(emptyCF, emptyCQ);
+            Result result = region.get(get);
+            if (result.isEmpty()) {
+                return true;
+            }
+            if (Bytes.compareTo(result.getValue(emptyCF, emptyCQ), 0, TRUE_BYTES.length,
+                    TRUE_BYTES, 0, TRUE_BYTES.length) == 0) {
+                return true;
+            }
+            return false;
+        }
+
+        private boolean verifyRowAndRemoveVerifiedCell(List<Cell> cellList) throws IOException {
+            boolean valid = true;
+            boolean verifiedColumnExist = false;
+            long cellListSize = cellList.size();
+            Cell cell = null;
+            if (cellListSize == 0) {
+                return true;
+            }
+            Iterator<Cell> cellIterator = cellList.iterator();
+            while (cellIterator.hasNext()) {
+                cell = cellIterator.next();
+
+                if (isVerifiedCell(cell)) {
+                    verifiedColumnExist = true;
+                    if (Bytes.compareTo(cell.getValueArray(), cell.getValueOffset(), cell.getValueLength(),
+                            TRUE_BYTES, 0, TRUE_BYTES.length) == 0) {
+                        valid = true;
+                    } else {
+                        valid = false;
+                    }
+                    if (cellListSize > 1) {
+                        cellIterator.remove();
+                    }
+                    break;
+                }
+            }
+            if (!verifiedColumnExist) {
+                byte[] rowKey = new byte[cell.getRowLength()];
+                System.arraycopy(cell.getRowArray(), cell.getRowOffset(), rowKey, 0, cell.getRowLength());
```

Comment:
We've had examples in the past of Cells that are not well-behaved when you try to access their fields and NPE. PHOENIX-5188 should take care of one major case, but good to be paranoid and put a try/catch here.

Reply:
If I catch an NPE, what to do with it? Should I log it and convert it to an IO exception?

Reply:
On second thought IndexedKeyValues only show up in the WAL, not when reading from a table, so this is fine.